### PR TITLE
Lazy serialization of layers

### PR DESCRIPTION
**Pull request by zhangyiatmicrosoft**
## Launch Checklist

this._serializedLayers dictionary is only needed when the style itself is serialized OR queryRenderedFeatures is called.

Currently it is populated upon creation, which is an overhead for initial map load. This PR changes to lazy serialization. 
Based on our experiment, for 150 layers this change saves ~20ms at 75 percentile, and ~100ms at 90 percentile.

Additional changes:

- "_serializeLayers" is renamed to "_serializeByIds" and piggyback on this._serializedLayers dictionary for better performance. Previously it was doing fresh serialization each time.
- "setState" method is serializing current instance twice. Fixed.
- replace forEach with simple for/of loop in "setState". Better performance, easier to debug
- creating layers is the most time-consuming part of _load method (~80% of time), so split out to be its own method for easier profiling in Chrome performance tool.

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Write tests for all new functionality.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

File: test/build/min.test.ts [hunk 1]
```diff
@@ -39,7 +39,7 @@ describe('test min build', () => {
         const decreaseQuota = 4096;
 
         // feel free to update this value after you've checked that it has changed on purpose :-)
-        const expectedBytes = 756547;
```
Comment (zhangyiatmicrosoft) [inline test/build/min.test.ts:42]: before this PR, the build size was already 757351.

File: src/style/style.ts [hunk 2]
```diff
@@ -441,12 +445,44 @@ class Style extends Evented {
...
-        for (const id of ids) {
-            const layer = this._layers[id];
-            if (layer.type !== 'custom') {
-                serializedLayers.push(layer.serialize());
+    /**
+     * take an array of string IDs, and based on this._layers, generate an array of LayerSpecification
+     * @param ids an array of string IDs, for which serilized layers will be generated. If ommited, all serialized layers will be returned
+     * @returns {Array<LayerSpecification>} generated result
+     */
+    private _serializeById(ids?: Array<string>): Array<LayerSpecification> {
```
Comment (HarelM) [inline src/style/style.ts:453]: _serializeByIds to indicate it's not for a single layer?

File: src/style/style.ts [hunk 3]
```diff
@@ -441,12 +445,44 @@ class Style extends Evented {
...
+        return serializedLayers;
+    }
+
+    /**
+     * Lazy initialization of this._serializedLayers dictionary and return it
+     * @returns this._serializedLayers dictionary
+     */
+    private _getSerializedLayers(): {[_: string]: LayerSpecification} {
+        let serializedLayers = this._serializedLayers;
+        if (!serializedLayers) {
```
Comment (HarelM) [inline src/style/style.ts:478]: Consider inverting this if to reduce indentation complexity.

File: src/style/style.test.ts [hunk 4]
```diff
@@ -570,6 +570,22 @@ describe('Style#_load', () => {
         // cleanup
         stub.mockReset();
     });
+
+    test('layers are NOT serialized immediately after creation', () => {
```
Comment (HarelM) [inline src/style/style.test.ts:574]: This test is great and important, but I would expect to see more tests here to cover the code changes that were made, for example, the "invalidation" of the layer serialization, update of data and check that the new serialization is up-to-date etc.

File: src/style/style.ts [hunk 5]
```diff
@@ -164,7 +164,7 @@ class Style extends Evented {
     _request: Cancelable;
     _spriteRequest: Cancelable;
     _layers: {[_: string]: StyleLayer};
-    _serializedLayers: {[_: string]: any};
+    _serializedLayers: {[_: string]: LayerSpecification};
```
Comment (HarelM) [inline src/style/style.ts:167]: Great!!! thanks!

File: src/style/style.ts [hunk 6]
```diff
@@ -206,8 +206,8 @@ class Style extends Evented {
 
         this._spritesImagesIds = {};
         this._layers = {};
-        this._serializedLayers = {};
-        this._order  = [];
+
+        this._order = [];
```
Comment (HarelM) [inline src/style/style.ts:210]: initialization to `null` would be a good practice here I think.

Comment (HarelM) [reply src/style/style.ts:210]: ahh, it is done in the below method. might be better to have it here though, IDK...

File: src/style/style.ts [hunk 7]
```diff
@@ -441,12 +445,44 @@ class Style extends Evented {
...
+    /**
+     * take an array of string IDs, and based on this._layers, generate an array of LayerSpecification
+     * @param ids an array of string IDs, for which serilized layers will be generated. If ommited, all serialized layers will be returned
+     * @returns {Array<LayerSpecification>} generated result
+     */
+    private _serializeById(ids?: Array<string>): Array<LayerSpecification> {
+
+        const serializedLayersDictionary = this._getSerializedLayers();
+        let serializedLayers: LayerSpecification[];
+        if (ids && ids.length > 0) {
```
Comment (HarelM) [inline src/style/style.ts:457]: consider inverting the if and returning `Object.values(serializedLayersDictionary)` in order to reduce indentation and complexity.

File: src/style/style.ts (hunk 2 above)
Comment (HarelM) [reply src/style/style.ts:453]: The method name is a bit confusing as when reading it you don't see any method that is called serialize so it's hard to understand where serialization is done....

File: src/style/style.ts [hunk 8]
```diff
@@ -441,12 +445,44 @@ class Style extends Evented {
...
+            serializedLayers  = Object.values(serializedLayersDictionary);
+        }
+        return serializedLayers;
+    }
+
+    /**
+     * Lazy initialization of this._serializedLayers dictionary and return it
+     * @returns this._serializedLayers dictionary
+     */
+    private _getSerializedLayers(): {[_: string]: LayerSpecification} {
```
Comment (HarelM) [inline src/style/style.ts:476]: See my comment above about the method name, it somewhat applied here as well.

File: src/style/style.ts [hunk 9]
```diff
@@ -894,7 +930,11 @@ class Style extends Evented {
         this._changed = true;
         this._removedLayers[id] = layer;
         delete this._layers[id];
-        delete this._serializedLayers[id];
+
+        const serializedLayers = this._serializedLayers;
```
Comment (HarelM) [inline src/style/style.ts:934]: I would consider removing this extra variable, I don't see it helps besides removing 6 characters twice.

File: src/style/style.ts [hunk 10]
```diff
@@ -1125,9 +1169,9 @@ class Style extends Evented {
             sprite: this.stylesheet.sprite,
             glyphs: this.stylesheet.glyphs,
             transition: this.stylesheet.transition,
-            sources: mapObject(this.sourceCaches, (source) => source.serialize()),
-            layers: this._serializeLayers(this._order)
-        }, (value) => { return value !== undefined; });
+            sources,
+            layers},
```
Comment (HarelM) [inline src/style/style.ts:1173]: move the `},` to the next line

File: src/style/style.ts [hunk 11]
```diff
@@ -1138,6 +1182,10 @@ class Style extends Evented {
             this._updatedSources[layer.source] = 'reload';
             this.sourceCaches[layer.source].pause();
         }
+
+        // upon updating, serilized layer dictionary should be reset.
+        // When needed, it will be populated with the correct copy again.
+        this._serializedLayers = null;
```
Comment (HarelM) [inline src/style/style.ts:1188]: Add a test for this scenario and a scenario for init -> update -> serialize or something like that.

**Comment by HarelM** [comment]
Thanks for taking the time to open this PR!
I've added a few comments.

File: src/style/style.ts (hunk 6 above)
Comment (zhangyiatmicrosoft) [reply src/style/style.ts:210]: _createLayers() will be called many times, so it has to set this._serializedLayers to null.
Adding another init statement here "feels" more complete but it is actually a waste. :)

File: src/style/style.ts (hunk 8 above)
Comment (zhangyiatmicrosoft) [reply src/style/style.ts:476]: renamed this one to _serializedAllLayers

File: src/style/style.ts (hunk 2 above)
Comment (zhangyiatmicrosoft) [reply src/style/style.ts:453]: added "s", since the other method is renamed, hope everything is easier to read.

File: src/style/style.ts (hunk 11 above)
Comment (zhangyiatmicrosoft) [reply src/style/style.ts:1191]: thank you. Added

File: src/style/style.test.ts (hunk 4 above)
Comment (zhangyiatmicrosoft) [reply src/style/style.test.ts:579]: done

File: src/style/style.ts [hunk 12]
```diff
@@ -1113,21 +1154,27 @@ class Style extends Evented {
     }
 
     serialize(): StyleSpecification {
+
+        const sources = mapObject(this.sourceCaches, (source) => source.serialize());
+        const layers = this._serializeByIds(this._order);
+        const myStyleSheet = this.stylesheet;
```
Comment (HarelM) [inline src/style/style.ts:1160]: Same here, although less important as there is no change to the state of the class.

File: src/style/style.ts [hunk 13]
```diff
@@ -441,14 +445,48 @@ class Style extends Evented {
...
+
+        return serializedLayers;
+    }
+
+    /**
+     * Lazy initialization of this._serializedLayers dictionary and return it
+     * @returns this._serializedLayers dictionary
+     */
+    private _serializedAllLayers(): {[_: string]: LayerSpecification} {
+        let serializedLayers = this._serializedLayers;
```
[the remark this answers is not in the document]
Comment (zhangyiatmicrosoft) [reply src/style/style.ts:476]: this.foo is never minimized, so to reduce runtime footprint as much as possible, it is a good practice to assign frequently used "this.foo" to a local var. In addition accessing local variable is a slightly faster.
Of course, everything is in terms of microseconds, but I take what I can get ...

File: src/style/style.ts (hunk 12 above)
Comment (zhangyiatmicrosoft) [reply src/style/style.ts:1160]: In this example, this.stylesheet is repeated 11 times, so we are talking about saving 110 bytes and 11 accesses to "this" object:
https://stackoverflow.com/questions/2936807/javascript-object-access-performance

Comment (HarelM) [reply src/style/style.ts:1160]: Thanks for this info! I did not know that.
Isn't there a typescript flag we can use to do this automatically?

**Review by HarelM — APPROVED** [review]

**Pull request merged**